### ISSUE TITLE
Yielding in Raylib Corrupt's Factor's OpenGL Context

### ISSUE THREAD

**Issue by Capital-EX** [factor/factor#2817]
When using Raylib, yielding Raylib's process cause Factor's UI to become corrupted in some way:

```
(U) [ set-namestack init-catchstack self quot>> call => stop ]
(O) update-ui-loop
(O) update-ui
(O) redraw-worlds
(O) draw-world
(O) (gl-error)
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
 ```

Minimum example that replicates the problem:

```factor
USING: calendar continuations raylib threads ;
IN: hello-raylib

MAIN: [
    600 450 "Corruption" init-window

    RAYWHITE clear-background begin-drawing 
        [ ] [ end-drawing ] finally 
    16 milliseconds sleep
    
    close-window
]
```

**Comment by Capital-EX** [comment]
This appears to happen when Raylib resumes after yield to Factor. Raylib is explicitly designed to operate under a single window model (https://github.com/raysan5/raylib/issues/2940). Doing a little testing `gl-function-context` when called in raylib first returns `f`, then we yield, finally when raylib resumes `gl-function-context` points to the current listener context. 

Given what I have been finding about how OpenGL context work, without explicitly switch back to raylib's context, Factor's UI gets clobbered. Can't really think of a good way to fix this, but I guess it's something more concrete than "weird bug is weird".

**Comment by Capital-EX** [comment]
Have to head to bed, but this example looks useful. 

https://github.com/raysan5/raylib/blob/b68d0850b1dee42d818e67d99f4affe1d890dba1/examples/others/rlgl_standalone.c#L160C22-L160C22

**Comment by Capital-EX** [comment]
I have solved it:

- Get the handle to factor's world context `world get-global handle>>`
- Switch back to it using `select-gl-context`
- Yield your current thread
- Get Raylibs window handle using `get-window-handle`
- Switch back to it using `glfwMakeContextCurrent`

```factor
: idle-frame ( -- ) 
    world get-global handle>> select-gl-context
    1000 milliseconds sleep
    [ "Success!" print ] \ run call-listener
    get-window-handle glfwMakeContextCurrent
;
```

Things to note: you must *never* yield in the middle of a `begin-drawing` call. Enabling this would require deep changes to how Factor handles concurrency. As every point a thread might sleep becomes a point of failure. 

Code outside of draw calls is free to yield as it please.

**Comment by Capital-EX** [comment]
A moment of victory: 
https://fedi.akselmo.dev/notice/AZb8VsjSvq3YiOVA3M

Need to fix find a fix for when the `world` variable changes from under the raylib side.

**Comment by Capital-EX** [comment]
Code I'm currently working with:
```factor
! Copyright (C) 2023 CapitalEx.
! See http://factorcode.org/license.txt for BSD license.
USING: accessors arrays calendar combinators.extras
combinators.smart continuations glfw.ffi io io.monitors
io.pathnames kernel literals namespaces opengl.gl.gtk
prettyprint random random.data raylib sequences threads timers
ui.backend ui.gadgets.worlds ui.tools.listener unicode
vocabs.loader ;
IN: hello-raylib

SYMBOL: color
SYMBOL: text
SYMBOL: rl-timers
CONSTANT: COLORS $[ RED GREEN BLUE 3array ]
CONSTANT: THIS-FILE $[ "vocab:hello-raylib/hello-raylib.factor" absolute-path ]

: current-color ( -- color ) 
    color get-global ;

: current-text ( -- text )
    text get-global ;

: clear ( -- )
    RAYWHITE clear-background ;

: hello-world ( -- ) 
    current-text 50 10 64 current-color draw-text ;

: idle-frame ( -- ) 
    world get-global handle>> select-gl-context
    16 milliseconds sleep
    get-window-handle glfwMakeContextCurrent
;

: with-draw ( quote -- )
    '[ begin-drawing clear @ ] [ end-drawing ] finally idle-frame ; inline

: randomize-color ( -- )
    COLORS random color set-global ;

: ranodmize-letter ( Letter --  Letter )
    [ Letter? ] [ drop random-Letter ] smart-when ;

: randomize-text ( -- )
    text [ [ ranodmize-letter ] map ] change-global ;

: start-monitor ( -- )
    [ THIS-FILE f [ drop [ "hello-raylib" reload ] forever ] with-monitor ] with-monitors ;

: ?hot-reloading ( ? -- )
    [ [ start-monitor ] "hot reloader" spawn drop ] when ;

: main ( -- )
    ! t ?hot-reloading

    600 450 "Hello, World!" init-window
    V{ } clone rl-timers set
        
    randomize-color
    "Hello, World!" text set-global

    [ randomize-color ] 250 milliseconds every rl-timers get push 
    [ randomize-text  ] 125  milliseconds every rl-timers get push

    [ window-should-close ] [ [ hello-world ] with-draw ] until
    
    close-window 
    
    rl-timers get [ stop-timer ] each
    f rl-timers set ;


MAIN: main
```

Causes the following error:
```factor
Generic word select-gl-context does not define a method for the POSTPONE: f class.
Dispatching on object: f

(U) [ set-namestack init-catchstack self quot>> call => stop ]
(O) main
(O) idle-frame
(O) M\ object select-gl-context
(O) no-method
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

When a window is closed, it's world is destroyed and it's thread is yielded. I think I just have to check if there is a world to restore a context to which should be a simple `when*`.

**Comment by greenfork** [comment]
> I have solved it:

This solution doesn't work as of current master, throws this error:

```
Generic word handle>> does not define a method for the POSTPONE: f class.
Dispatching on object: f

(U) [ c-to-factor => ]
    c-to-factor
(U) [ [ (get-catchstack) push ] dip call => (get-catchstack) pop* ]
(O) command-line-startup
(O) run-script
(O) main
(O) idle-frame
(O) M\ object handle>>
(O) no-method
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

I'm on x86_64 Linux, here is code for reproduction:

```factor
USING: calendar kernel math namespaces prettyprint raylib
threads io glfw.ffi ;
IN: raylib-demo

CONSTANT: screen-width 800
CONSTANT: screen-height 640

SYMBOL: cnt

: dbg ( obj -- ) unparse 10 10 16 BLACK draw-text ;

: make-window ( -- )
    screen-width screen-height "Raylib Demo" init-window
    1 set-target-fps ;

: clear-window ( -- ) RAYWHITE clear-background ;

: idle-frame ( -- )
    world get-global handle>> select-gl-context
    1000 milliseconds sleep
    [ "Success!" print ] \ run call-listener
    get-window-handle glfwMakeContextCurrent
;

: game-loop ( -- )
    cnt get dbg
    cnt get .
    cnt get 1 + cnt set
    begin-drawing
    clear-window
    end-drawing
    idle-frame
    ; inline

: main ( -- )
    0 cnt set
    make-window
    [ game-loop window-should-close not ] loop
    close-window ;

MAIN: main
```

**Comment by mrjbq7** [comment]
So, this is going to be ``f`` when only using raylib to ``init-window``:

```world get-global```

So perhaps to prevent an error it would be something like:

```factor
world get-global [ handle>> select-gl-context ] when*
```

**Comment by greenfork** [comment]
I tried it. When running from terminal as a script, it spews this error after around 5 seconds:

```
Error in thread 40 (Listener call, [ ~quotation~ \ run ~listener-gadget~ dup wait-for-listener...):

timed-out-error
timer T{ timer f ~curried~ 5000000000 f f f ~thread~ }

(U) [ set-namestack init-catchstack self quot>> call => stop ]
(U) [ dup wait-for-listener => [ (call-listener) ] with-ctrl-break ]
(O) wait-for-flag-timeout
(O) timed-out-error
(O) M\ object throw
(U) [
        OBJ-CURRENT-THREAD special-object error-thread set-global
        current-continuation => error-continuation set-global
        [ original-error set-global ] [ rethrow ] bi
    ]
```

Via listener, it also returns an error and spoils UI after a couple of seconds:

```
An error occurred while drawing the world T{ world f ~array~ ~array~ f f ~vector~ ~array~....
This world has been deactivated to prevent cascading errors.
Memory protection fault at address 0x58
```

I'm a bit hopeless to resolve this.

On hot reloading: traditional compiled languages can do a hot-code reloading by separating the game into main exe which never changes, and a dll with the main game logic. Then this dll can be dynamically loaded and re-loaded with `dlopen`. Not ideal, but still a thing. If Factor can do dlls, this can be accomplished as well.

**Comment by mrjbq7** [comment]
Well the timeout for wait-for-listener makes sense it doesn't work without a listener.  The author of this issue had posted a few demos doing live hot reloading with raylib and factor, for example:

https://www.youtube.com/watch?v=IphW1Gbqh58

**Comment by greenfork** [comment]
I found out with printf debugging that the problem happens after the second iteration in the `end-drawing` function.

The videos are a good demonstration but they don't show how to implement it. I have browsed through the author's public repositories, no success there too.

**Comment by greenfork** [comment]
Commenting out these lines from `extra/glfw/ffi/ffi.factor` removes the crashes.

```factor
<<
"glfw" {
    { [ os windows? ] [ "glfw3.dll" ] }
    { [ os macos? ] [ "glfw3.dylib" ] }
    { [ os unix? ] [ "libglfw3.so" ] }
} cond cdecl add-library
>>
```

The GLFW is statically compiled in Raylib. I assume that when I link GLFW externally, there's either a version mismatch and it doesn't work, or GLFW has an internal state that is kept inside the one compiled into Raylib.

I still don't know how to yield control to the listener because it only responds after I close the Raylib window.

**Comment by Capital-EX** [comment]
I saw you comment and yeah, it's been a while. The tricks I was doing with glfw don't seem to work. Mostly because I can't figure out how to get factor to load glfw on Arch Linux.

The last demo I showed was using this code:
```factor
: with-ui-safe-yield ( quot -- )
    world get-global handle>> [ select-gl-context ] when*
        call
    get-window-handle glfwMakeContextCurrent ; inline

: idle-frame ( -- ) 
   [ 16 milliseconds sleep ] with-ui-safe-yield ; inline

: with-draw ( quote -- )
    '[ begin-drawing clear @ ]
     [ end-drawing ] finally idle-frame ; inline
    
: until-window-should-close ( quote -- ) 
    f @@died set-global
    [ window-should-close @@died get-global or ] 
    swap 
    '[ _ 
        [ 
            '[ _ 
                {
                    { "Resume Raylib" f }
                    { "Close Raylib"  t }
                } rethrow-restarts @@died set-global
            ] with-ui-safe-yield
        ] recover 
    ] until ; inline
    ```
    
I was using factor's error handling to recover errors from raylib and restore factor's glcontext. But, I'm not sure any of this works any more.

**Comment by greenfork** [comment]
Great, thank you! I will try to make it into a vocabulary. I got basic things working, now I have a better idea about error handling

**Comment by greenfork** [comment]
@Capital-EX did you run the game loop in another thread? Or you managed to unblock the listener on the main thread where the game runs?

**Comment by Capital-EX** [comment]
@greenfork I run the game loop in a background thread. `sleep` will pause the current thread, so you have to spin up new one. I usually used `[ run-game ] spawn`.

**Comment by greenfork** [comment]
I think I solved the problem with calling into listener directly via `call-listener`

**Comment by greenfork** [comment]
I think this issue can be closed as fixed in https://github.com/factor/factor/pull/3078

The direct solution is to reset OpenGL context when not interacting with the game window, can be achieved by either a) swapping the OpenGL context to the one listener uses; or b) using the GLFW library that is statically compiled into Raylib to do so. I like the option (b) more because it does not assume anything about interaction of different systems: Factor Listener and Raylib. The code to execute any quotation in the Listener context:

```factor
: with-listener-context ( quot -- )
    f make-context-current
    call
    get-window-handle make-context-current ; inline
```

where `make-context-current` is from GLFW and `get-window-handle` is from Raylib. See the above PR for the concrete implementation and usage.

**Comment by mrjbq7** [comment]
Awesome, thanks for all the contributions!
